File: scala/ch02/ex2.42.scala

def enumerateInterval (start: Int, end: Int) = {
  def iter (result:List[Int], n:Int): List[Int] = 
    if (n > end)
      result
    else
      iter (result ++ List(n), n + 1)
  iter(Nil, start)
}

type Position = List[Int]

val emptyBoard:List[Position] = List(Nil)

// Represent each position as a list of rows. The position in
// the list is the column, in reverse, with both rows and columns counting from 1.
def adjoinPosition (newRow: Int, col:Int, restOfQueens:List[Position]) =
  restOfQueens map ((position: Position) => newRow :: position)

// By definition, this is the only queen in the k'th column, so we just
// check rows and diagonals for previous columns.
def safe (k:Int, position: List[Position]) = {
  val p = position.head
  val newRow = position.head.head
  def checkRows (pos: Position): Boolean = pos match {
    case Nil => true
    case _ if newRow == pos.head => false
    case _ => checkRows (pos.tail)
  }
  def checkDiags (n:Int, pos: Position): Boolean = pos match {
    case Nil => true
    case _ if (pos.head == newRow + n) || (pos.head == newRow - n) => false
    case _ => checkDiags (n+1, pos.tail)
  }
  checkRows (p.tail) && checkDiags (1, p.tail)
}

def queens (boardSize: Int) = {
  def queenCols (k: Int):List[List[Position]] = 
    if (k == 0)
      List(emptyBoard)
    else
      // use Scala's flatMap now.
      queenCols (k - 1) flatMap {restOfQueens: List[Position] =>
        enumerateInterval(1, boardSize) map {
          newRow:Int => adjoinPosition (newRow, k, restOfQueens)
        }
      } filter { position: List[Position] => safe (k, position) }

  queenCols (boardSize)
}

import org.scalatest._ 
import org.scalatest.matchers._

object queensSpec extends Spec with ShouldMatchers {
  def testQueen (n: Int, expectedN: Int) = {
    val queensN = queens(n)
    format("board size = %d, # solutions = %d: %s\n", n, queensN.size, queensN)
    queensN.size should equal (expectedN)
  }
  
  describe ("queens") {
    it ("should return the correct number of correct answers for a given board size") {
      Map(0 -> 1, 1 -> 1, 2 -> 0, 3 -> 0, 4 -> 2, 
          5 -> 10, 6 -> 4, 7 -> 40, 8 -> 92).foreach { (ne:Pair[Int,Int]) =>
        testQueen (ne._1, ne._2)
      }
    }
  }
}
queensSpec execute

File: clojure/ch02/ex2.42.clj
(ns sicp.ch02 (:use clojure.test))

; Using vectors because they work better.

; Recall that accumulate is really fold-right
(defn fold-right [op initial sequence]
  (if (empty? sequence)
      initial
      (op (first sequence)
          (fold-right op initial (vec (rest sequence))))))

(defn flatmap [proc seq]
  (vec (fold-right into [] (vec (map proc seq)))))
  
(defn enumerate-interval [start end]
  (seq (loop [result [] n start]
        (if (> n end)
        result
        (recur (conj result n) (+ n 1))))))

(def empty-board [[]])

; Represent each position as a vector of rows. The position in
; the vector is the column, with both rows and columns counting from 1.
(defn adjoin-position [new-row col rest-of-queens]
  (map (fn [position] (conj position new-row)) rest-of-queens))

; By definition, this is the only queen in the k'th column, so we just
; check rows and diagonals for previous columns.
(defn safe? [k position]
  (let [p (first position) new-row (last p)]
    (defn check-row [i]
      (cond (= i k) true
            (let [i-row (nth p (- i 1)) delta (- k i)]
            ;(println "k, i, delta, i-row, new-row = " k, i delta i-row new-row)
              (or (= i-row new-row) (= i-row (+ new-row delta)) (= i-row (- new-row delta)))) false
            :else (check-row (+ i 1)))))
    (check-row 1))

(defn queens [board-size]
  (defn queen-cols [k]
    (if (= k 0)
        [empty-board]
        (vec (filter 
          (fn [position] (safe? k position))
          (flatmap
            (fn [rest-of-queens]
              (map (fn [new-row]
                     (adjoin-position new-row k rest-of-queens))
                   (enumerate-interval 1 board-size)))
            (queen-cols (- k 1)))))))
    (queen-cols board-size))

(defn length [coll]
  (loop [result 0 c coll]
    (if (empty? c) result
        (recur (+ result 1) (rest c)))))

(loop [n 0]
  (let [queens-n (queens n) len-queens-n (length queens-n)]
    (println "board size = " n " # solutions = " len-queens-n ": " queens-n))
  (if (< n 8)
      (recur (+ n 1))))

(deftest test-queens
  (is (= 1  (length (queens 0)))) ; actually, it's zero, but we have an empty [] placeholder.
  (is (= 1  (length (queens 1))))
  (is (= 0  (length (queens 2))))
  (is (= 0  (length (queens 3))))
  (is (= 2  (length (queens 4))))
  (is (= 10 (length (queens 5))))
  (is (= 4  (length (queens 6))))
  (is (= 40 (length (queens 7))))
  (is (= 92 (length (queens 8)))))
  
(run-tests)
